server/src/test/java/org/elasticsearch/indices/recovery/MultiFileWriterTests.java

```diff
@@ -0,0 +1,113 @@
+        directorySpy = spy(directory);
+        store = createStore(indexShard.shardId(), indexShard.indexSettings(), directorySpy);
+    }
+
+    @Override
+    public void tearDown() throws Exception {
+        super.tearDown();
+        directory.close();
+        closeShards(indexShard);
+    }
+
+    public void testWritesFile() throws IOException {
+        var fileWriter = createMultiFileWriter(true);
+        var file = createFile("file");
+
+        fileWriter.writeFile(file.metadata, 10, new ByteArrayInputStream(file.bytes));
+
+        verify(directorySpy).createOutput("temp_file", IOContext.DEFAULT);
+        verify(directorySpy).sync(Collections.singleton("temp_file"));
+    }
+
+    public void testFailsToWriteFileWithIncorrectChecksum() throws IOException {
+        var fileWriter = createMultiFileWriter(true);
+        var file = createFile("file");
+
+        expectThrows(
+            CorruptIndexException.class,
+            () -> fileWriter.writeFile(withWrongChecksum(file.metadata), 10, new ByteArrayInputStream(file.bytes))
+        );
+
+        verify(directorySpy).createOutput("temp_file", IOContext.DEFAULT);
+        verify(directorySpy).deleteFile("temp_file");
+        verify(directorySpy, never()).sync(anyCollection());
+    }
+
+    public void testWritesFileWithIncorrectChecksumWithoutVerification() throws IOException {
+        var fileWriter = createMultiFileWriter(false);
+        var file = createFile("file");
+
+        fileWriter.writeFile(withWrongChecksum(file.metadata), 10, new ByteArrayInputStream(file.bytes));
+
+        verify(directorySpy).createOutput("temp_file", IOContext.DEFAULT);
+        verify(directorySpy).sync(Collections.singleton("temp_file"));
+    }
+
+    private MultiFileWriter createMultiFileWriter(boolean verifyOutput) {
+        return new MultiFileWriter(store, mock(RecoveryState.Index.class), "temp_", logger, mock(Runnable.class), verifyOutput);
+    }
+
+    private record FileAndMetadata(byte[] bytes, StoreFileMetadata metadata) {}
+
+    private static FileAndMetadata createFile(String name) throws IOException {
+        var buffer = new ByteBuffersDataOutput();
+        var output = new ByteBuffersIndexOutput(buffer, "test", name);
+        output.writeString(TestUtil.randomRealisticUnicodeString(random(), 10, 1024));
+        CodecUtil.writeBEInt(output, CodecUtil.FOOTER_MAGIC);
+        CodecUtil.writeBEInt(output, 0);
+        String checksum = Store.digestToString(output.getChecksum());
+        CodecUtil.writeBELong(output, output.getChecksum());
+        output.close();
+
+        return new FileAndMetadata(buffer.toArrayCopy(), new StoreFileMetadata(name, buffer.size(), checksum, Version.CURRENT.toString()));
+    }
+
+    private static StoreFileMetadata withWrongChecksum(StoreFileMetadata metadata) {
+        return new StoreFileMetadata(metadata.name(), metadata.length(), randomAlphaOfLength(6), metadata.writtenBy());
```

Comment:
maybe use randomValueOtherThan() to ensure the `randomAlphaOfLength(6)` is never equal to the metadata checksum?